pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -77,16 +82,83 @@ func NewClusterMemberRemovalController(
 			masterNodeInformer,
 			masterMachineInformer,
 			kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Informer(),
+			kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Informer(),
 		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
 }
 
 func (c *clusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
-	// stop if the machine API is not functional
+	// only attempt to scale down if the machine API is functional
 	if isFunctional, err := c.machineAPIChecker.IsFunctional(); err != nil {
 		return err
 	} else if !isFunctional {
 		return nil
 	}
+
+	var errs []error
+	if err := c.removeMemberWithoutMachine(ctx); err != nil {
+		errs = append(errs, err)
+	}
+	if err := c.attemptToScaleDown(ctx); err != nil {
+		errs = append(errs, err)
+	}
+	if err := c.attemptToRemoveLearningMember(ctx); err != nil {
+		errs = append(errs, err)
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+// attemptToRemoveLearningMember attempts to remove a learning member pending deletion regardless of whether a replacement member has been found
+func (c *clusterMemberRemovalController) attemptToRemoveLearningMember(ctx context.Context) error {
+	currentVotingMemberIPListSet, err := c.votingMemberIPListSet()
+	if err != nil {
+		return err
+	}
+	memberMachines, err := c.currentMemberMachines()
+	if err != nil {
+		return err
+	}
+	var learningMachines []*machinev1beta1.Machine
+	for memberMachineIP, memberMachine := range ceohelpers.IndexMachinesByNodeInternalIP(memberMachines) {
+		if !currentVotingMemberIPListSet.Has(memberMachineIP) {
+			learningMachines = append(learningMachines, memberMachine)
+		}
+	}
+	return c.removeMemberPendingDeletion(ctx, learningMachines, "learning")
+}
+
+// attemptToScaleDown attempts to remove a voting member only once we have identified that
+// a Machine resource is being deleted and a replacement member has been created
+func (c *clusterMemberRemovalController) attemptToScaleDown(ctx context.Context) error {
+	currentVotingMemberIPListSet, err := c.votingMemberIPListSet()
+	if err != nil {
+		return err
+	}
+
+	desiredControlPlaneReplicasCount, err := ceohelpers.ReadDesiredControlPlaneReplicasCount(c.configMapListerForKubeSystemNamespace)
```

Comment:
/cc @JoelSpeed I think in the KEP we wanted to get this info from the `cluster-config-v1` configmap in `kube-system` namespace

Reply:
> We need a different signal for this. Install-config is write-once data. It cannot be modified via normal means, lacks a schema, lacks validation, and should not be used to feed post-install operator behavior.

does OCP support changing the number of nodes post installation?

Reply:
> We need a different signal for this. Install-config is write-once data. It cannot be modified via normal means, lacks a schema, lacks validation, and should not be used to feed post-install operator behavior.

The etcd operator has been relying on this value to know how many replicas it thinks it should have all the way through OpenShift 4. This is the only place in the cluster today where the value of the control plane size actually resides at present.

> It would be best to get this from something like the control plane machine set resource. But even something written into the etcd.operator.openshift.io schema could work. Add it to spec, then observe it from there.

This was written down somewhere in an enhancement, the plan eventually is that the ControlPlaneMachineSet becomes the source of truth for the replica count, but, this feature is going to be delivered before ControlPlaneMachineSet (which is looking likely to miss 4.11), so the plan was to fall back to the old mechanism until CPMS becomes available

> I think in the KEP we wanted to get this info from the cluster-config-v1 configmap in kube-system namespace

Yep, this was discussed with Sam and Lili before they left, it's the only source of the control plane size that we have right now, in the future that will change but we do need to be able to handle upgrades and clusters that won't have CPMS ever 

> does OCP support changing the number of nodes post installation?

Today, no. All customers are required to have 3 replicas in their control plane _unless_ they are granted a support exception, in which case they are allowed to have 5, but that is from install time only. In the future, we want to look into horizontal scaling of the control plane, but this is an explicit non-goal of the first pass of the etcd protection and CPMS projects

Reply:
thanks for the input.